File: source code/mf-mc.py
from collections import deque

def bfs(residual_graph, source, sink, parent):
    visited = [False] * len(residual_graph)
    queue = deque([source])
    visited[source] = True
    parent[source] = -1

    while queue:
        u = queue.popleft()

        for v in range(len(residual_graph)):
            if not visited[v] and residual_graph[u][v] > 0:
                queue.append(v)
                visited[v] = True
                parent[v] = u
                if v == sink:
                    return True

    return False

def edmonds_karp(graph, source, sink):
    n = len(graph)
    residual_graph = [[0] * n for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            residual_graph[i][j] = graph[i][j]

    parent = [-1] * n
    max_flow = 0

    while bfs(residual_graph, source, sink, parent):
        path_flow = float('inf')
        v = sink

        while v != source:
            u = parent[v]
            path_flow = min(path_flow, residual_graph[u][v])
            v = u

        v = sink
        while v != source:
            u = parent[v]
            residual_graph[u][v] -= path_flow
            residual_graph[v][u] += path_flow
            v = u

        max_flow += path_flow

    return max_flow, residual_graph

def dfs(graph, u, visited):
    visited[u] = True
    for v in range(len(graph)):
        if graph[u][v] > 0 and not visited[v]:
            dfs(graph, v, visited)

def min_cut(graph, source, sink, input_file, output_file):
    max_flow, residual_graph = edmonds_karp(graph, source, sink)

    output_file.write("Flow Network Graphical Representation:\n")
    for i in range(len(graph)):
        for j in range(len(graph)):
            if graph[i][j] > 0:
                output_file.write(f"{i} -> {j} (Capacity: {graph[i][j]})\n")
    output_file.write("\n")

    visited = [False] * len(graph)
    dfs(graph, source, visited)
    output_file.write("Vertices on the source side of the minimum cut:\n")
    for i in range(len(visited)):
        if visited[i]:
            output_file.write(f"{i}\n")

    for i in range(len(graph)):
        for j in range(len(graph)):
            if visited[i] and not visited[j] and graph[i][j] > 0:
                output_file.write(f"{i} -> {j} (Capacity: {graph[i][j]})\n")

    output_file.write("\n")
    output_file.write(f"Edges in the Minimum Cut: {sum(visited)}\n")
    output_file.write(f"Maximum Flow: {max_flow}\n")

def main():
    input_file = open("input.txt", "r")
    output_file = open("output.txt", "w")

    source, sink = map(int, input_file.readline().split())
    graph = [list(map(int, input_file.readline().split())) for _ in range(6)]

    min_cut(graph, source, sink, input_file, output_file)

    input_file.close()
    output_file.close()

if __name__ == "__main__":
    main()

File: source code/mf-mc.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

#define V 6 

bool bfs(vector<vector<int>>& residualGraph, int source, int sink, vector<int>& parent) {
    vector<bool> visited(V, false);
    queue<int> q;
    q.push(source);
    visited[source] = true;
    parent[source] = -1;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v = 0; v < V; ++v) {
            if (!visited[v] && residualGraph[u][v] > 0) {
                q.push(v);
                visited[v] = true;
                parent[v] = u;
                if (v == sink)
                    return true;
            }
        }
    }

    return false;
}

int edmondsKarp(vector<vector<int>>& graph, int source, int sink) {
    vector<vector<int>> residualGraph = graph; 
    vector<int> parent(V, -1);
    int maxFlow = 0;

    while (bfs(residualGraph, source, sink, parent)) {
        int pathFlow = INT_MAX;
        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            pathFlow = min(pathFlow, residualGraph[u][v]);
        }

        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            residualGraph[u][v] -= pathFlow;
            residualGraph[v][u] += pathFlow;
        }

        maxFlow += pathFlow;
    }

    return maxFlow;
}

void dfs(vector<vector<int>>& residualGraph, int u, vector<bool>& visited) {
    visited[u] = true;
    for (int v = 0; v < V; ++v) {
        if (residualGraph[u][v] > 0 && !visited[v]) {
            dfs(residualGraph, v, visited);
        }
    }
}

void minCut(vector<vector<int>>& graph, int source, int sink, ofstream& outputFile) {
    vector<vector<int>> residualGraph = graph;
    int maxFlow = edmondsKarp(graph, source, sink);
    vector<bool> visited(V, false);

    dfs(residualGraph, source, visited);

    outputFile << "Flow Network Graphical Representation:" << endl;
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
            if (graph[u][v] > 0) {
                outputFile << u << " -> " << v << " (Capacity: " << graph[u][v] << ")" << endl;
            }
        }
    }
    outputFile << endl;

    int t=0;
    outputFile << "Vertices on the source side of the minimum cut: " << endl;
    for (int u = 0; u < V; ++u) {
        if (visited[u]) {
            outputFile << u << endl;
            t++;
        }
    }
    outputFile << "Edges in the Minimum Cut:" << t<< endl;
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
                if (visited[u] && !visited[v] && graph[u][v] > 0) {
                outputFile << u << " -> " << v << " (Capacity: " << graph[u][v] << ")" << endl;
            }
        }
    }
    outputFile << endl;
    outputFile << "Maximum Flow: " << maxFlow << endl;
}

int main() {
    ifstream inputFile("input.txt");
    ofstream outputFile("output.txt");
    vector<vector<int>> graph(V, vector<int>(V, 0));
    int source, sink;
    inputFile >> source >> sink;
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            inputFile >> graph[i][j];
        }
    }
    minCut(graph, source, sink, outputFile);
    inputFile.close();
    outputFile.close();

    return 0;
}
